### PR TITLE
[WIP][JIT] Better error message when trying to instantiate NamedTuple

### PR DESCRIPTION
**Pull request by jamesr66a**

### CODE REVIEW

File: torch/csrc/jit/script/python_sugared_value.cpp [hunk 1]
```diff
@@ -509,6 +511,13 @@ std::shared_ptr<SugaredValue> toSugaredValue(
     if (auto classType = pyCu.get_class(c10::QualifiedName(qualifiedName))) {
       return std::make_shared<ClassValue>(classType);
     }
+    // Use a heuristic here to identify NamedTuple instances:
+    // 1) must be a subclass of tuple
+    // 2) Has an attribute "_fields"
+    auto tuple_type = (PyObject*)&((&PyTuple_Type)->ob_base);
+    if (PyObject_IsSubclass(obj.ptr(), tuple_type) && py::hasattr(obj, "_fields")) {
```
Comment (driazati) [inline torch/csrc/jit/script/python_sugared_value.cpp]: What's wrong with `py::isinstance(obj, py::tuple)`?

**Comment by jamesr66a** [comment]
@driazati obj is not an instance but rather the class type itself. I tried that at first and it always returned false because of that

Sent from my Samsung Smart Fridge

On Jun 3, 2019, at 3:13 PM, driazati <notifications@github.com<mailto:notifications@github.com>> wrote:


@driazati commented on this pull request.

________________________________

In torch/csrc/jit/script/python_sugared_value.cpp<https://github.com/pytorch/pytorch/pull/21309#discussion_r290059375>:

> @@ -509,6 +511,13 @@ std::shared_ptr<SugaredValue> toSugaredValue(
     if (auto classType = pyCu.get_class(c10::QualifiedName(qualifiedName))) {
       return std::make_shared<ClassValue>(classType);
     }
+    // Use a heuristic here to identify NamedTuple instances:
+    // 1) must be a subclass of tuple
+    // 2) Has an attribute "_fields"
+    auto tuple_type = (PyObject*)&((&PyTuple_Type)->ob_base);
+    if (PyObject_IsSubclass(obj.ptr(), tuple_type) && py::hasattr(obj, "_fields")) {


What's wrong with py::isinstance(obj, py::tuple)?

**Comment by driazati** [comment]
I'd rather catch the error in `annotations.py` or whatever and just use `issubclass` instead of CPython madness

**Comment by jamesr66a** [comment]
@driazati I do it this way because it intercepts the issue earlier and prevents the construction of the bogus SugaredValue for the NamedTuple. We're going to have to go this route down the road anyway when we add first-class support for this stuff, so I think it makes sense to have it here

**Review by driazati — APPROVED** [review]

**Pull request closed without merging**